CombinatorialGames/Game/Computability/FGame.lean

```diff
@@ -0,0 +1,418 @@
+
+theorem isOption_wf : WellFounded IsOption := by
+  refine ⟨rec fun s t f g IHl IHr ↦ .intro _ ?_⟩
+  rintro y (h | h)
+  · exact IHl _
+  · exact IHr _
+
+instance : WellFoundedRelation SGame := ⟨_, isOption_wf⟩
+
+macro "sgame_wf" : tactic =>
+  `(tactic| all_goals solve_by_elim
+    [Prod.Lex.left, Prod.Lex.right, PSigma.Lex.left, PSigma.Lex.right,
+    IsOption.moveLeft, IsOption.moveRight] )
+
+instance instDecidableEq : ∀ x y : SGame, Decidable (x = y)
+  | mk m n f g, mk m' n' f' g' => if h : m = m' ∧ n = n' then by
+    have : ∀ a b, Decidable (f a = f' b) := fun a b ↦ instDecidableEq ..
+    have : ∀ a b, Decidable (g a = g' b) := fun a b ↦ instDecidableEq ..
+    rw [mk.injEq, Fin.heq_fun_iff h.1, Fin.heq_fun_iff h.2]
+    infer_instance
+  else .isFalse (by simp_all)
+
+/-- The identical relation on short games. See `SGame.Identical`. -/
+def Identical : SGame.{u} → SGame.{u} → Prop
+  | mk _ _ xL xR, mk _ _ yL yR =>
+      Relator.BiTotal (fun i j ↦ Identical (xL i) (yL j)) ∧
+      Relator.BiTotal (fun i j ↦ Identical (xR i) (yR j))
+
+@[inherit_doc] scoped infix:50 " ≡ " => SGame.Identical
+
+theorem identical_iff : ∀ {x y : SGame}, x ≡ y ↔
+    Relator.BiTotal (x.moveLeft · ≡ y.moveLeft ·) ∧ Relator.BiTotal (x.moveRight · ≡ y.moveRight ·)
+  | mk .., mk .. => Iff.rfl
+
+@[refl]
+protected theorem Identical.refl (x) : x ≡ x :=
+  x.recOn fun _ _ _ _ IHL IHR ↦ ⟨Relator.BiTotal.refl IHL, Relator.BiTotal.refl IHR⟩
+
+@[symm]
+protected theorem Identical.symm : ∀ {x y}, x ≡ y → y ≡ x
+  | mk .., mk .., ⟨hL, hR⟩ => ⟨hL.symm fun _ _ h ↦ h.symm, hR.symm fun _ _ h ↦ h.symm⟩
+
+@[trans]
+protected theorem Identical.trans : ∀ {x y z}, x ≡ y → y ≡ z → x ≡ z
+  | mk .., mk .., mk .., ⟨hL₁, hR₁⟩, ⟨hL₂, hR₂⟩ =>
+    ⟨hL₁.trans (fun _ _ _ h₁ ↦ h₁.trans) hL₂, hR₁.trans (fun _ _ _ h₁ ↦ h₁.trans) hR₂⟩
+
+/-- `Identical` as a `Setoid`. -/
+def identicalSetoid : Setoid SGame :=
+  ⟨Identical, .refl, .symm, .trans⟩
+
+/-- If `x ≡ y`, then a left move of `x` is identical to some left move of `y`. -/
+theorem Identical.moveLeft : ∀ {x y}, x ≡ y → ∀ i, ∃ j, x.moveLeft i ≡ y.moveLeft j
+  | mk .., mk .., ⟨hl, _⟩ => hl.1
+
+/-- If `x ≡ y`, then a left move of `y` is identical to some left move of `x`. -/
+theorem Identical.moveLeft_symm : ∀ {x y}, x ≡ y → ∀ i, ∃ j, x.moveLeft j ≡ y.moveLeft i
+  | mk .., mk .., ⟨hl, _⟩ => hl.2
+
+/-- If `x ≡ y`, then a right move of `x` is identical to some right move of `y`. -/
+theorem Identical.moveRight : ∀ {x y}, x ≡ y → ∀ i, ∃ j, x.moveRight i ≡ y.moveRight j
+  | mk .., mk .., ⟨_, hr⟩ => hr.1
+
+/-- If `x ≡ y`, then a right move of `y` is identical to some right move of `x`. -/
+theorem Identical.moveRight_symm : ∀ {x y}, x ≡ y → ∀ i, ∃ j, x.moveRight j ≡ y.moveRight i
+  | mk .., mk .., ⟨_, hr⟩ => hr.2
+
+instance Identical.instDecidable (a b) : Decidable (a ≡ b) := by
+  have : DecidableRel (a.moveLeft · ≡ b.moveLeft ·) := fun c d ↦ Identical.instDecidable ..
```

Comment:
Since this builds data, shouldn't it be:
```suggestion
  let : DecidableRel (a.moveLeft · ≡ b.moveLeft ·) := fun c d ↦ Identical.instDecidable ..
```